packages/@aws-cdk/aws-rds/lib/instance-parameter-group.ts

```diff
@@ -0,0 +1,134 @@
+import { CfnOutput, Construct, IConstruct, Token } from '@aws-cdk/cdk';
+import { Parameters } from './props';
+import { CfnDBParameterGroup } from './rds.generated';
+
+/**
+ * A instance parameter group
+ */
+export interface IInstanceParameterGroup extends IConstruct {
+  /**
+   * Name of this parameter group
+   */
+  readonly parameterGroupName: string;
+
+  /**
+   * Export this parameter group
+   */
+  export(): InstanceParameterGroupImportProps;
+}
+
+/**
+ * Properties to reference a instance parameter group
+ */
+export interface InstanceParameterGroupImportProps {
+  readonly parameterGroupName: string;
+}
+
+/**
+ * Properties for a instance parameter group
+ */
+export interface InstanceParameterGroupProps {
+  /**
+   * Database family of this parameter group
+   *
+   * @see https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_CreateDBParameterGroup.html#API_CreateDBParameterGroup_RequestParameters
+   * @example
+   * { family: 'MySQL5.5' }
+   */
+  family: string;
+
+  /**
+   * Description for this parameter group
+   */
+  description: string;
+
+  /**
+   * The parameters in this parameter group
+   */
+  parameters: Parameters;
+}
+
+/**
+ * Defina a instance parameter group
+ */
+export class InstanceParameterGroup extends Construct implements IInstanceParameterGroup {
+  /**
+   * Import a parameter group
+   */
+  public static import(scope: Construct, id: string, props: InstanceParameterGroupImportProps): IInstanceParameterGroup {
+    return new ImportedInstanceParameterGroup(scope, id, props);
+  }
+
+  public readonly parameterGroupName: string;
+  private readonly parameters: Parameters = {};
+
+  constructor(scope: Construct, id: string, props: InstanceParameterGroupProps) {
+    super(scope, id);
+
+    const { description, family, parameters } = props;
+
+    const resource = new CfnDBParameterGroup(this, 'Resource', {
+      description,
+      family,
+      parameters: new Token(() => this.parameters)
+    });
+
+    for (const [key, value] of Object.entries(parameters || {})) {
+      this.setParameter(key, value);
+    }
+
+    this.parameterGroupName = resource.dbParameterGroupName;
+  }
+
+  /**
+   * Export this parameter group
+   */
+  public export(): InstanceParameterGroupImportProps {
+    return {
+      parameterGroupName: new CfnOutput(this, 'ParameterGroupName', { value: this.parameterGroupName }).makeImportValue().toString(),
+    };
+  }
+
+  /**
+   * Set a single parameter in this parameter group
+   */
+  public setParameter(key: string, value: string | undefined) {
+    if (value === undefined && key in this.parameters) {
+      delete this.parameters[key];
```

Comment:
Don't overload the `setParameter` function with delete behavior. We don't support union types in public APIs because they don't translate well to other languages like Java, and it is also a confusing API to call `set` to delete something.

`setParameter` is for setting parameters and `removeParameter` is for removing them, keep it simple.

Reply:
@sam-goodwin I think this PR is based on the existing implementation of the cluster parameter group (https://github.com/awslabs/aws-cdk/blob/master/packages/%40aws-cdk/aws-rds/lib/cluster-parameter-group.ts implemented here https://github.com/awslabs/aws-cdk/pull/729) which proposes the same approach.

Does you comment mean that the cluster parameter group should also be refactored?

I'm currently working on a PR to support db instances and would like to have your opinion on this.

Reply:
On a second pass, I realize my comment about this union type was incorrect - `string | undefined` will work fine because it means optional. My mistake.

That said, it does seem confusing to me that we use `setParameter` to delete a parameter. I wouldn't expect to (for example) use `map.put(key, null)` to delete a *key* from a `Map` in Java. I was unaware of the precedent set in cluster parameter groups. What is the use case for this overloaded behavior?

Reply:
I don't know, it seems to come from this discussion https://github.com/awslabs/aws-cdk/pull/729#discussion_r218038746.

For me it doesn't make sense to have `setParameter` and `removeParameter` methods, `parameters` should be defined directly in the construction props.

Reply:
I agree. @rix0rrr can you chime in?

Reply:
I think it might make sense to build the RDS parameters object incrementally, since it can be *frking* huge (there are many parameters to set!), so I think I might want some function to do:

```
lessDurabilityMorePerformance(paramGroup);
```

Or something. But I don't feel super strongly about this, I'm okay with removing the mutators as well.